datafusion/physical-expr/src/expressions/case.rs

```diff
@@ -86,13 +108,35 @@ impl CaseExpr {
         when_then_expr: Vec<WhenThen>,
         else_expr: Option<Arc<dyn PhysicalExpr>>,
     ) -> Result<Self> {
+        // normalize null literals to None in the else_expr (this already happens
+        // during SQL planning, but not necessarily for other use cases)
+        let else_expr = match &else_expr {
+            Some(e) => match e.as_any().downcast_ref::<Literal>() {
+                Some(lit) if lit.value().is_null() => None,
+                _ => else_expr,
+            },
+            _ => else_expr,
+        };
+
         if when_then_expr.is_empty() {
             exec_err!("There must be at least one WHEN clause")
         } else {
+            let eval_method = if expr.is_some() {
+                EvalMethod::WithExpression
+            } else if when_then_expr.len() == 1
```

Comment:
I was thinking about this optimization and I think it would be valid for any CASE expression that had a NULL ELSE, not just column

So in other words, I think you could remove the `when_then_expr[0].1.as_any().is::<Column>()` check and this would still work fine

Reply:
Generally, it is only safe to use this approach for expressions that are infallible. One of the main reasons that regular CaseExpr is expensive is that we need to only evaluate the "true" expression on rows where the predicate has evaluated to true.

I do think that this could be extended beyond just Column expressions though.

Reply:
I will add a comment in the code to explain this

Reply:
I added more comments to explain the rationale and what is safe or not